### PR TITLE
Switch to linkify-react for element Linkification as it handles React subtrees without exploding (#10060

### PR DESCRIPTION
**Pull request by t3chguy**
Notes: Fix a series of soft crashes where errors include "The node to be removed is not a child of this node."

More specifically, if the subtree which `linkifyElement` (including via our homebrew `Linkify` component) was called upon contained other React Elements this would cause a soft crash during unmounting operations. In the SpaceHierarchy case this was when a topic contained URLs as those were in their own spans if the room was being joined. It was also possible on the RoomTopic component. `linkify-react` properly handles react subtrees.

Fixes https://github.com/vector-im/element-web/issues/22752
Regressed by https://github.com/matrix-org/matrix-react-sdk/commit/0d100cbbce0d3ea4b3c8e9503befd1a5837503c2

Bypassing coverage requirement in favour of Cypress regression test

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix a series of soft crashes where errors include "The node to be removed is not a child of this node." ([\#10060](https://github.com/matrix-org/matrix-react-sdk/pull/10060)). Fixes vector-im/element-web#22752.

### CODE REVIEW

File: cypress/e2e/spaces/spaces.spec.ts [hunk 1]
```diff
@@ -281,4 +283,29 @@ describe("Spaces", () => {
...
+                preset: "public_chat" as Preset,
+                name: "Test Room",
+                topic: "This is a topic https://github.com/matrix-org/matrix-react-sdk/pull/10060 with a link",
+            });
+            const { room_id: spaceId } = await bot.createRoom(spaceCreateOptions("Test Space", [roomId]));
+            await bot.invite(spaceId, user.userId);
+        });
+
+        cy.getSpacePanelButton("Test Space").should("exist");
+        cy.wait(500); // without this we can end up clicking too quickly and it ends up having no effect
```
Comment (andybalaam) [inline cypress/e2e/spaces/spaces.spec.ts:299]: Is there something we can wait for here instead of just sleeping?

File: src/components/structures/SpaceHierarchy.tsx [hunk 2]
```diff
@@ -210,6 +210,25 @@ const Tile: React.FC<ITileProps> = ({
...
     }
 
+    let topicSection: ReactNode | undefined;
+    if (topic) {
+        topicSection = (
+            <Linkify
+                options={{
+                    attributes: {
+                        onClick(ev: MouseEvent) {
+                            // prevent clicks on links from bubbling up to the room tile
```
Comment (andybalaam) [inline src/components/structures/SpaceHierarchy.tsx:220]: The old code appears to check for A tags specifically and here we don't - is that a deliberate change?

File: src/components/structures/SpaceHierarchy.tsx [hunk 3]
```diff
@@ -210,6 +210,25 @@ const Tile: React.FC<ITileProps> = ({
...
+                            ev.stopPropagation();
+                        },
+                    },
+                }}
+            >
+                {" · "}
```
Comment (andybalaam) [inline src/components/structures/SpaceHierarchy.tsx:226]: ~~Previously we had a dot only if there was no topic but here we always add one. Is that a deliberate change?~~

Nope. Previously the code confused me and now it's clearer :-)

File: src/components/views/elements/RoomTopic.tsx [hunk 4]
```diff
@@ -71,12 +70,14 @@ export default function RoomTopic({ room, ...props }: IProps): JSX.Element {
...
                         <Linkify
-                            as="p"
-                            onClick={(ev: MouseEvent) => {
-                                if ((ev.target as HTMLElement).tagName.toUpperCase() === "A") {
-                                    modal.close();
-                                }
+                            options={{
+                                attributes: {
+                                    onClick() {
+                                        modal.close();
```
Comment (andybalaam) [inline src/components/views/elements/RoomTopic.tsx:76]: Same question about A tags

File: cypress/e2e/spaces/spaces.spec.ts [hunk 5]
```diff
@@ -32,7 +33,7 @@ function openSpaceContextMenu(spaceName: string): Chainable<JQuery> {
     return cy.get(".mx_SpacePanel_contextMenu");
 }
 
-function spaceCreateOptions(spaceName: string): ICreateRoomOpts {
+function spaceCreateOptions(spaceName: string, roomIds: string[] = []): ICreateRoomOpts {
```
Comment (andybalaam) [inline cypress/e2e/spaces/spaces.spec.ts:36]: Is this change related to the rest or a small extra fix?

File: src/SlashCommands.tsx [hunk 6]
```diff
@@ -501,14 +501,11 @@ export const Commands = [
...
-            const ref = (e): void => {
-                if (e) linkifyElement(e);
-            };
-            const body = topicToHtml(topic.text, topic.html, ref, true);
+            const body = topicToHtml(topic.text, topic.html, undefined, true);
 
             Modal.createDialog(InfoDialog, {
                 title: room.name,
-                description: <div ref={ref}>{body}</div>,
+                description: <Linkify>{body}</Linkify>,
```
Comment (andybalaam) [inline src/SlashCommands.tsx:508]: This is awesome. So much better!

**Review by andybalaam — APPROVED** [review]
A few questions, nothing blocking. Thanks!

File: cypress/e2e/spaces/spaces.spec.ts (hunk 1 above)
Comment (t3chguy) [reply cypress/e2e/spaces/spaces.spec.ts:299]: I tried alternatives but nothing seemed reliable enough at least locally, this is likely a symptom of a race/bug but not one that needs to be tackled as part of fixing this huge regression

File: src/components/structures/SpaceHierarchy.tsx (hunk 2 above)
Comment (t3chguy) [reply src/components/structures/SpaceHierarchy.tsx:220]: The old code bound a handler on the parent div, this binds the handler on all the linkified elements (anchor tags)

File: src/components/views/elements/RoomTopic.tsx (hunk 4 above)
Comment (t3chguy) [reply src/components/views/elements/RoomTopic.tsx:76]: The old code bound a handler on the parent div, this binds the handler on all the linkified elements (anchor tags)

File: cypress/e2e/spaces/spaces.spec.ts (hunk 5 above)
Comment (t3chguy) [reply cypress/e2e/spaces/spaces.spec.ts:36]: This is extending a test utility to be able to create a space with a room in it to streamline test fixture setup
